zipline/pipeline/loaders/events.py

```
@@ -0,0 +1,154 @@
+        occurred along with other associated values.
+
+        If the DataFrames contain a "timestamp" column, that column is
+        interpreted as the date on which we learned about the event.
+
+        If the DataFrames do not contain a "timestamp" column, we assume we
+        knew about the event on all prior dates.  This mode is only supported
+        if ``infer_timestamp`` is explicitly passed as a truthy value.
+    infer_timestamps : bool, optional
+        Whether to allow omitting the "timestamp" column.
+    dataset : DataSet
+        The DataSet object for which this loader loads data.
+    expected_cols : frozenset
+        Set of expected columns for the dataset, without timestamp.
+    """
+
+    def __init__(self,
+                 all_dates,
+                 events_by_sid,
+                 infer_timestamps=False,
+                 dataset=None,
+                 expected_cols=frozenset()):
+        self.all_dates = all_dates
+        # Do not modify the original in place, since it may be used for other
+        #  purposes.
+        self.events_by_sid = (
+            events_by_sid.copy()
+        )
+        dates = self.all_dates.values
+
+        for k, v in iteritems(events_by_sid):
+            # First, must convert to DataFrame.
+            if isinstance(v, pd.Series):
+                # If Series was passed, DateTime index is assumed.
+                self.events_by_sid[k] = pd.DataFrame(v)
+            elif isinstance(v, pd.DatetimeIndex):
+                if not infer_timestamps:
+                    raise ValueError(
+                        "Got DatetimeIndex for sid %d.\n"
+                        "Pass `infer_timestamps=True` to use the first date in"
+                        " `all_dates` as implicit timestamp." % k
+                    )
+                self.events_by_sid[k] = pd.DataFrame(v)
+                v.index = [dates[0]] * len(v)
+            # Already a DataFrame
+            elif isinstance(v, pd.DataFrame):
```

I think this should not be an elif. If we have hit either of the two branches above then we will have promoted to a dataframe but we still need to make these checks. Also, as written now we will silently ignore anythin that is not a datetimeindex, series, or dataframe.

Can we add some tests for these conversion cases?